### PR TITLE
Replace minFlash with flushDuration

### PR DESCRIPTION
**Pull request by zimin2000**
Currently there is an issue that a low frequency report can delay reporting for a very long time (minute+). This PR is replacing `minFlush` with  `flushDelay` that unconditionally flushes the buffer after `flushDelay` time passed since first report was buffered.

### CODE REVIEW

File: gostc.go [hunk 1]
```diff
@@ -77,22 +77,22 @@ func NewClient(addr string) (*Client, error) {
...
 //
-// The three parameters queueSize, maxPacketBytes, and minFlush tune the
-// buffered channel size, maximum single packet size, and minimum time between
-// flushes. Message are buffered until maxPacketBytes is reached or until some
-// time as passed (no more than minFlush). Use NewDefaultBufferedClient for
-// reasonable defaults.
+// The three parameters queueSize, maxPacketBytes, and flushDelay tune the
+// buffered channel size, maximum single packet size, and the flush delay.
+// Messages are buffered until maxPacketBytes is reached or until flushDelay
+// time passed since first message was buffered. Use NewDefaultBufferedClient
```
Comment (cespare) [inline gostc.go]: > time has passed since the first message was buffered.

File: gostc.go [hunk 2]
```diff
@@ -135,15 +135,16 @@ func (c *Client) bufferAndSend(maxPacketBytes int, minFlush time.Duration) {
 				}
 				buf = buf[:0]
 			}
-			timer.Reset(minFlush)
 		case msg := <-c.incoming:
 			if len(buf)+len(msg)+1 > maxPacketBytes {
 				c.send(buf)
 				buf = buf[:0]
-				timer.Reset(minFlush)
+				timer.Stop()
```
Comment (cespare) [inline gostc.go]: Stop races with the timer firing, so if we stop it here we should check whether it fired anyway and drain the channel if so.

```
if !timer.Stop() {
	<-timer.C
}
```

See the Stop documentation for more details.

File: gostc.go [hunk 3]
```diff
@@ -105,25 +105,25 @@ const (
...
 
 func (c *Client) send(b []byte) error {
 	_, err := c.c.Write(b)
 	return err
 }
 
-func (c *Client) bufferAndSend(maxPacketBytes int, minFlush time.Duration) {
-	timer := time.NewTimer(minFlush)
+func (c *Client) bufferAndSend(maxPacketBytes int, flushDelay time.Duration) {
+	timer := time.NewTimer(flushDelay)
```
Comment (cespare) [inline gostc.go]: I think the initial state should be for the timer to be stopped. That matches the behavior inside the loop. Then you can document the invariant:

```
// This timer is stopped whenever len(buf) == 0.
timer := time.NewTimer(0)
<-timer.C
```

File: gostc.go [hunk 4]
```diff
@@ -135,15 +135,16 @@ func (c *Client) bufferAndSend(maxPacketBytes int, minFlush time.Duration) {
...
 		case msg := <-c.incoming:
 			if len(buf)+len(msg)+1 > maxPacketBytes {
 				c.send(buf)
 				buf = buf[:0]
-				timer.Reset(minFlush)
+				timer.Stop()
 			}
 			if len(buf) > 0 {
 				buf = append(buf, '\n')
+			} else {
```
Comment (cespare) [inline gostc.go]: Very minor, but I think this loop now reads better the other way around:

```
if len(buf) == 0 { ... } else { ... }
```

File: gostc_test.go [hunk 5]
```diff
@@ -159,23 +158,15 @@ func TestBufferedMaxSize(t *testing.T) {
...
-	// Assigning flushHook here isn't racy because the the bufferAndSend
-	// goroutine doesn't access flushHook unless there are messages in its
-	// buffer, and we haven't sent any.
-	c.flushHook = func() { once.Do(func() { close(ch) }) }
 	for i := byte(0); i < 4; i++ {
 		c.Set("a", []byte{'a' + i})
-		if i == 1 {
-			<-ch // wait for flush
-		}
+		time.Sleep(7*time.Millisecond)
```
Comment (cespare) [inline gostc_test.go]: Why did you change the way this test works? It was intentionally written to not be timing-dependent. That's a very nice property for the avoidance of flakes and to make tests be fast.

**Review by cespare — CHANGES_REQUESTED** [review]

File: gostc_test.go (hunk 5 above)
Comment (zimin2000) [reply gostc_test.go]: Rolled back.

**Comment by cespare** [comment]
I merged this with some minor tweaks.
